### PR TITLE
Ensure custody of timezones from query to results, add debugging info

### PR DESCRIPTION
**Pull request by senior**
This PR makes several changes intended to fix timezone related
issues and expose more information when there are problems. For
databases that support report timezones, this commit fixes an issue
where the JVM's timezone was used to interpret datetimes that didn't
include a timezone. The report timezone should be used instead. This
commit also adds a step to the sync process to detect and store the
databases timezone, which will be helpful in troubleshooting timezone
issues with a new timezone debugging page (to be added later).

### CODE REVIEW

File: src/metabase/driver/generic_sql/query_processor.clj [hunk 1]
```diff
@@ -383,31 +420,8 @@
...
-                        (assert (re-matches #"[A-Za-z\/_]+" <>)))
-        format-string (sql/set-timezone-sql driver)
-        sql           (format format-string (str \' timezone \'))]
-    (log/debug (u/format-color 'green "Setting timezone with statement: %s" sql))
-    (jdbc/db-do-prepared connection [sql])))
-
-(defn- run-query-without-timezone [driver settings connection query]
-  (do-in-transaction connection (partial run-query query)))
-
-(defn- run-query-with-timezone [driver settings connection query]
```
Comment (salsakran) [inline src/metabase/driver/generic_sql/query_processor.clj]: Dumb question, but does this affect date extracts (which should be done with respect to the reporting timezone)?

File: test/metabase/sync/sync_metadata/sync_timezone_test.clj [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  (:require  [expectations :refer :all]
+             [toucan.util.test :as tt]
+             [metabase.models
+              [database :refer [Database]]
+              [table :refer [Table]]]
+             [toucan.db :as db]
+             [metabase.sync.sync-metadata.sync-timezone :as sync-tz]
+             [metabase.test.data.datasets :as datasets]
+             [metabase.test.data :as data]
+             [clj-time.core :as time]
```
Comment (camsaul) [inline test/metabase/sync/sync_metadata/sync_timezone_test.clj]: ns sorting 😉

File: test/metabase/sync/sync_metadata/sync_timezone_test.clj [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+             [metabase.util :as u]))
+
+(defn- db-timezone [db-or-id]
+  (db/select-one-field :timezone Database :id (u/get-id db-or-id)))
+
+(datasets/expect-with-engines #{:h2 :postgres}
```
Comment (camsaul) [inline test/metabase/sync/sync_metadata/sync_timezone_test.clj]: what exactly are we testing here? A quick comment would be nice since otherwise it's not really clear what's going on

File: src/metabase/driver.clj [hunk 4]
```diff
@@ -203,7 +227,8 @@
    :table-rows-seq                    (fn [driver & _]
                                         (throw
                                          (NoSuchMethodException.
-                                          (str (name driver) " does not implement table-rows-seq."))))})
+                                          (str (name driver) " does not implement table-rows-seq."))))
+   :current-db-time                   (constantly nil)})
```
Comment (salsakran) [inline src/metabase/driver.clj]: is this time or timezone?

File: test/metabase/query_processor_test/date_bucketing_test.clj [hunk 5]
```diff
@@ -226,27 +294,41 @@
...
+
+    (or (oracle? *engine*)
+        (contains? #{:sqlserver :sqlite :crate} *engine*))
     [[23 54] [24 46] [25 39] [26 61]]
 
+    (supports-report-timezone? *engine*)
+    [[23 49] [24 47] [25 39] [26 58] [27 7]]
+
+
+
```
Comment (camsaul) [inline test/metabase/query_processor_test/date_bucketing_test.clj]: why so many blank lines?

File: src/metabase/driver/presto.clj [hunk 6]
```diff
@@ -40,25 +42,34 @@
 
 (defn- parse-time-with-tz [s]
   ;; Try parsing with offset first then with full ZoneId
+  (println "getting ready to parse " s)
```
Comment (salsakran) [inline src/metabase/driver/presto.clj]: I think we'd want to remove this before merging.

File: src/metabase/driver/presto.clj [hunk 7]
```diff
@@ -40,25 +42,34 @@
...
   (or (u/ignore-exceptions (u/parse-date "HH:mm:ss.SSS ZZ" s))
       (u/parse-date "HH:mm:ss.SSS ZZZ" s)))
 
 (defn- parse-timestamp-with-tz [s]
   ;; Try parsing with offset first then with full ZoneId
+  (println "getting ready to parse " s)
```
Comment (salsakran) [inline src/metabase/driver/presto.clj]: I think we'd want to remove this before merging.

File: test/metabase/driver/postgres_test.clj [hunk 8]
```diff
@@ -7,9 +7,7 @@
              [query-processor-test :refer [rows]]
              [sync :as sync]
              [util :as u]]
-            [metabase.driver
-             [generic-sql :as sql]
-             postgres]
```
Comment (camsaul) [inline test/metabase/driver/postgres_test.clj]: `cljr-clean-namespace` probably did this for you automatically but the reason we're loading the postgres driver here is that the `import` statement below it will fail if some other namespace hasn't already loaded the postgres driver. Thus something like `lein test metabase.driver.postgres-test` could potentially fail.

It's unfortunate that there's no way to tell `cljr-clean-ns` that some of those lines are there on purpose like that. But I would just keep that in mind when you review your own PR, if there's a required namespace that doesn't look like it's being used for something being removed take a look at the `:import`s as well

File: src/metabase/util.clj [hunk 9]
```diff
@@ -858,3 +858,37 @@
...
+  (->DateTimeFormatter "yyyy-MM-dd HH:mm:ss.SSS"))
+
+(def ^:private ordered-date-parsers
+  "When using clj-time.format/parse without a formatter, it tries all
+  default formatters, but not ordered by how likely the date
+  formatters will succeed. This leads to very slow parsing as many
+  attempts fail before the right one is found. Using this retains that
+  flexibility but improves performance by trying the most likely ones
+  first"
+  (let [most-likely-default-formatters [:mysql :date-hour-minute-second :date-time :date
```
Comment (camsaul) [inline src/metabase/util.clj]: 👍

**Review by salsakran — APPROVED** [review]
Minor comments, but I would really like a little more description of what is being tested. Just changing the bin counts makes it hard to follow. 

It would be especially useful to somehow explicitly call out reporting timezone, database timezone, and jvm timezone interactions in the the test suite.

File: src/metabase/driver/presto.clj (hunk 6 above)
Comment (senior) [inline src/metabase/driver/presto.clj]: Could have swore I got rid of those. Good catch, I'll remove them

File: test/metabase/driver/postgres_test.clj (hunk 8 above)
Comment (senior) [inline test/metabase/driver/postgres_test.clj]: Yeah, I've been having issues with that, running cljr-clean-ns, then it breaking the code. I'll add this back.

File: src/metabase/sync/fetch_metadata.clj [hunk 10]
```diff
@@ -22,3 +23,8 @@
   (let [driver (driver/->driver database)]
     (when (driver/driver-supports? driver :foreign-keys)
       (driver/describe-table-fks driver database table))))
+
+(s/defn ^:always-validate db-timezone :- s/Str
```
Comment (camsaul) [inline src/metabase/sync/fetch_metadata.clj]: It might be possible to define an output schema a little stricter than just `s/Str` here. You could do `su/NonBlankString` or even better define a new schema that checks whether the timezone is legit. You could do something like check whether `(java.util.TimeZone/getTimeZone <str>)` works correctly (for an invalid timezone it just returns `GMT` so you could check whether the `TimeZone` you get back isn't GMT, unless of course the string passed in was `"GMT"`, in which case that would be ok)

File: src/metabase/driver/presto.clj [hunk 11]
```diff
@@ -18,7 +18,9 @@
             [metabase.query-processor.util :as qputil]
             [metabase.util
              [honeysql-extensions :as hx]
-             [ssh :as ssh]])
+             [ssh :as ssh]]
+            [clj-time.format :as tformat]
+            [clj-time.core :as time])
```
Comment (camsaul) [inline src/metabase/driver/presto.clj]: ns sorting

File: src/metabase/driver/presto.clj (hunk 6 above)
Comment (camsaul) [inline src/metabase/driver/presto.clj]: assuming this is left over from debugging? You can leave it there just make it `log/debug` or something

File: src/metabase/driver/presto.clj (hunk 7 above)
Comment (camsaul) [inline src/metabase/driver/presto.clj]: same thing, `log/debug`

File: test/metabase/driver/postgres_test.clj (hunk 8 above)
Comment (camsaul) [inline test/metabase/driver/postgres_test.clj]: Yeah I would just try to review all the changes in your code when you're wrapping up a PR and look for things like that and fix as needed. `cljr-clean-ns` is still better than doing things yourself since now it's a manual review of an automatic process instead of a manual review of a manual process

**Comment by senior** [comment]
@salsakran A description would have been useful for me too as I was making these changes. I'll do some research and write up some comments around the test.

As far as assertions around timezones in the test suite, I've added a test for each database asserting the timezone that the database is expected to be in for the test. With the date tests (date bucketing and unix timestamp) I've added an explicit reporting timezone (pacific). I could add an assertion for the JVM timezone to the date bucketing tests to cover that. Is that the kind of thing you were looking for?

**Comment by salsakran** [comment]
I was thinking along the lines of:
if the jvm timezone is Central, the db timezone is pacific and the reporting timezone is eastern, for these 3 rows grouped by day, are all the day counts correct.

**Review by camsaul — APPROVED** [review]
LGTM

**Comment by mazameli** [comment]
@senior, does this also resolve:
#3642 
#3744 
#4180
#4245
#2517
#4372
and #2035?

Those issues were all noted as being dupes according to @salsakran, and 2035 as being the "root cause." Should these be closed as well?

**Comment by senior** [comment]
#3642 - Looking at @tlrobinson's comment (here)[https://github.com/metabase/metabase/issues/3642#issuecomment-265856616] it looks like it's an issue on the UI side, query results that contain multiple offsets, such as a daylight savings shift in the middle of a resultset

#3744 - Fixed with this PR

#4180 not sure, my guess would be not as the druid stuff goes through a very different code path, worth testing though

#4245 - I'll validate this, it looks like the correct data is coming back, but being displayed incorrectly

#2517 - Fixed with this PR

#4372 - No but I put some advice on addressing this through what we have today

#2035 - Fixed with this PR
